src/data/program_configuration.js

```diff
@@ -299,8 +388,171 @@ class CompositeExpressionBinder<T> implements Binder<T> {
+
+        const zoomInArray = this.zoomInPaintVertexArray;
+        const zoomOutArray = this.zoomOutPaintVertexArray;
+
+        const start = zoomInArray.length;
+
+        zoomInArray.reserve(length);
+        zoomOutArray.reserve(length);
+
+        const min = this.expression.evaluate({zoom: this.zoom - 1}, feature);
+        const mid = this.expression.evaluate({zoom: this.zoom }, feature);
+        const max = this.expression.evaluate({zoom: this.zoom + 1}, feature);
+        if (imagePositions) {
+            const imageMin = imagePositions[min];
+            const imageMid = imagePositions[mid];
+            const imageMax = imagePositions[max];
+
+            if (!imageMin || !imageMid || !imageMax) return;
+
+            for (let i = start; i < length; i++) {
+                zoomInArray.emplaceBack(
+                    imageMid.tl[0], imageMid.tl[1], imageMid.br[0], imageMid.br[1],
+                    imageMin.tl[0], imageMin.tl[1], imageMin.br[0], imageMin.br[1]
+                );
+
+                zoomOutArray.emplaceBack(
+                    imageMid.tl[0], imageMid.tl[1], imageMid.br[0], imageMid.br[1],
+                    imageMax.tl[0], imageMax.tl[1], imageMax.br[0], imageMax.br[1]
+                );
+            }
+        }
     }
+
+    updatePaintArray(start: number, end: number, feature: Feature, featureState: FeatureState, imagePositions: {[string]: ImagePosition}) {
+        // We populate two paint arrays because, for cross-faded properties, we don't know which direction
+        // we're cross-fading to at layout time. In order to keep vertex attributes to a minimum and not pass
+        // unnecessary vertex data to the shaders, we determine which to upload at draw time.
+
+        const zoomInArray = this.zoomInPaintVertexArray;
+        const zoomOutArray = this.zoomOutPaintVertexArray;
+
+        const min = this.expression.evaluate({zoom: this.zoom - 1}, feature, featureState);
+        const mid = this.expression.evaluate({zoom: this.zoom }, feature, featureState);
+        const max = this.expression.evaluate({zoom: this.zoom + 1}, feature, featureState);
+
+        if (imagePositions) {
+            const imageMin = imagePositions[min];
+            const imageMid = imagePositions[mid];
+            const imageMax = imagePositions[max];
+
+            if (!imageMin || !imageMid || !imageMax) return;
+            for (let i = start; i < end; i++) {
+                zoomInArray.emplace(i,
+                    imageMid.tl[0], imageMid.tl[1], imageMid.br[0], imageMid.br[1],
+                    imageMin.tl[0], imageMin.tl[1], imageMin.br[0], imageMin.br[1]
+                );
+
+                zoomOutArray.emplace(i,
+                    imageMid.tl[0], imageMid.tl[1], imageMid.br[0], imageMid.br[1],
+                    imageMax.tl[0], imageMax.tl[1], imageMax.br[0], imageMax.br[1]
+                );
+            }
+        }
+    }
+
+    getVertexBuffer(crossfade: CrossfadeParameters) {
```

Comment:
I think it would be clearer to inline this into `updatePatternPaintBuffers` below -- do we expect to use it anywhere else?